src/rpc/server.cpp

```
@@ -370,14 +368,12 @@ static const CRPCCommand vRPCCommands[] =
     { "rawtransactions",    "fundrawtransaction",     &fundrawtransaction,     false },
 #endif
 
-#ifdef ENABLE_ADDRESS_INDEXING
     /* Address index */
     { "addressindex",       "getaddressmempool",      &getaddressmempool,      true  },
```

If used on a node that is not started with `-addressindex`, I believe `getaddressmempool` would produce no error, but return an empty response.
I'd suggest to be more explicit and throw an error, as it happens for the other RPCs. To be consistent with other RPCs, a good place to add a check and throw an error might be `CTxMemPool::getAddressIndex()`.

---

getaddressmempool returns an error

```
    if (!mempool.getAddressIndex(addresses, indexes)) {
        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available for address");
    }
```

---

True, but `CTxMemPool::getAddressIndex()` always returns `true`. Indeed, my suggestion was to make `CTxMemPool::getAddressIndex()` return `false` if `-addressindex` is not enabled.

---

Sorry @drgora if `CTxMemPool::getAddressIndex()` always return true (which I can confirm) what is the point of having this ?

```
    if (!mempool.getAddressIndex(addresses, indexes)) {
        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available for address");
    }
```
This gets never executed then

---

I agree with you, this is never executed as of today.
One way of fixing this might be removing these code lines altogether, but I think that it would be better to actually throw this error if addressindex is not enabled. In other words, I believe that throwing an explicit RPC error is way more clear than returning an empty result.

---

The quickest way I believe is to check for fAddressIndex in CTxMemPool::getAddressIndex and return false if the first is false.
And secondly to change the check in RPC call as 

```
    if (!mempool.getAddressIndex(addresses, indexes)) {
        throw JSONRPCError(RPC_INTERNAL_ERROR, "Address indexing not enabled");
    } else if (indexes.empty()) {}
        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available for address");
    }
```
WDYT ?

---

I fully agree on the first point, that's actually what I meant to suggest, sorry if it was not clear.

On the second, I'm not sure. I'd say that an empty result is not an error per se (if addressindex is enabled), so we might even avoid throwing an error in that case. But I don't have strong opinions on this, your suggestion is also ok for me.

---

@drgora ok ... give a second look at df63ccf

---

Lgtm, thanks